### PR TITLE
Add an experimental local database URL check for the test configuration

### PR DESCRIPTION
**Pull request by alarochelle**
In order to prevent mistakes when running tests in an application, we add an experimental local database URL check in the `test.exs` configuration file.

If the specified `DATABASE_URL` used to run tests is not a local URL, an error will be raised, thus preventing the tests from running and from potentially damaging the database.

This change is experimental in the sense that it may not be _necessary_, but could be very useful in some cases.

### CODE REVIEW

File: config/releases.exs [hunk 1]
```diff
@@ -31,6 +31,24 @@ defmodule Environment do
...
+      %{host: "localhost"} ->
+        url
+
+      %{host: "127.0.0.1"} ->
+        url
+
+      %{host: nil} ->
+        url
+
+      %{host: _} ->
```
Comment (remi) [inline config/releases.exs]: Since we’re using this pattern as a catch-all, we could to:

```suggestion
      _ ->
```

File: config/releases.exs [hunk 2]
```diff
@@ -31,6 +31,24 @@ defmodule Environment do
...
+        raise "Make sure that the DATABASE_URL environment variable for the tests points to a local database."
```
Comment (Boubalou) [inline config/releases.exs]: The method look re-usable for any URL but the error it raises is bound to DATABASE_URL.

File: config/releases.exs [hunk 3]
```diff
@@ -31,6 +31,24 @@ defmodule Environment do
       _ -> nil
     end
   end
+
+  def get_local_url(key) do
```
Comment (simonprev) [inline config/releases.exs]: I think the function could be a bit smaller and the error message more helpful:

```elixir
url = get(key)

case URI.parse(url).host do
  host when host in ~w(localhost 127.0.0.1) -> url
  host -> raise "Expected host of the DATABASE_URL environment variable to be localhost or 127.0.0.1, got: #{host}"
end
```

File: config/releases.exs [hunk 4]
```diff
@@ -31,6 +31,24 @@ defmodule Environment do
...
+    url = get(key)
+
+    case URI.parse(get(key)) do
```
Comment (remi) [inline config/releases.exs]: Since we already got it:

```suggestion
    case URI.parse(url) do
```

File: config/releases.exs (hunk 2 above)
Comment (Boubalou) [reply config/releases.exs]: Maybe explicitly rename it to `get_database_url()`, what do you think?

Comment (remi) [inline config/releases.exs]: I think we just need to rephrase the message and make it a bit more generic:

```suggestion
        raise "Make sure that the #{key} environment variable points to a local URL."
```

File: config/releases.exs (hunk 3 above)
Comment (remi) [reply config/releases.exs]: If we’re going with this, I suggest

```elixir
  host -> raise "Expected host of the #{key} environment variable to be `localhost` or `127.0.0.1`, got: #{host}"
```

**Comment by Boubalou** [comment]
I agree with the initiative 💯 !

File: config/releases.exs (hunk 2 above)
Comment (alarochelle) [reply config/releases.exs]: I think we should go with what @simonprev and @remi proposed which is to make the error message more generic by using the `key` and `host` parameters.

Comment (Boubalou) [reply config/releases.exs]: Yes!

File: config/releases.exs [hunk 5]
```diff
@@ -31,6 +31,15 @@ defmodule Environment do
       _ -> nil
     end
   end
+
+  def get_local_url(key) do
+    url = get(key)
+
+    case URI.parse(url).host do
+      host when host in ~w(localhost 127.0.0.1) -> url
```
Comment (charlesdemers) [inline config/releases.exs]: Should we allow `0.0.0.0` and `host.docker.internal`?

Comment (alarochelle) [reply config/releases.exs]: You're right, they should probably be allowed as there is no risk of doing so anyways. What do others think about this?

File: config/releases.exs [hunk 6]
```diff
@@ -31,6 +31,15 @@ defmodule Environment do
       _ -> nil
     end
   end
+
+  def get_local_url(key) do
+    url = get(key)
+
+    case URI.parse(url).host do
+      host when host in ~w(0.0.0.0 127.0.0.1 localhost host.docker.internal) -> url
+      host -> raise "Expected host of the #{key} environment variable to be `localhost` or `127.0.0.1`, got: #{host}"
```
Comment (romarickb) [inline config/releases.exs]: Maybe we could use a varaible for  `~w(0.0.0.0 127.0.0.1 localhost host.docker.internal)` and use it to be more precise in the raise message exception

Comment (remi) [reply config/releases.exs]: True, it’s a little misleading to provide two valid values when we actually have four valid ones 😄

**Pull request merged**